lib/sidekiq/job/iterable.rb

```diff
@@ -0,0 +1,261 @@
+      #
+      #   sidekiq_options iteration: { retry_backoff: 30 } # in seconds
+      #
+      #  or globally:
+      #
+      #   Sidekiq::Config::DEFAULTS[:iteration][:retry_backoff] = 30 # in seconds
+      #
+      # @return [Boolean]
+      #
+      def throttle?
+        false
+      end
+
+      # @api private
+      def perform(*arguments)
+        @arguments = arguments
+        fetch_previous_iteration_state
+
+        @executions += 1
+        @start_time = Time.now
+
+        enumerator = build_enumerator(*arguments, cursor: @cursor)
+        unless enumerator
+          logger.debug("'#build_enumerator' returned nil, skipping the job.")
+          return
+        end
+
+        assert_enumerator!(enumerator)
+
+        if @executions == 1
+          on_start
+        else
+          on_resume
+        end
+
+        completed = catch(:abort) do
+          iterate_with_enumerator(enumerator, arguments)
+        end
+
+        on_shutdown
+        completed = handle_completed(completed)
+
+        if completed
+          on_complete
+
+          logger.info {
+            format("Completed iterating. times_interrupted=%d total_time=%.3f", @times_interrupted, @total_time)
+          }
+        else
+          reenqueue_iteration_job
+        end
+      end
+
+      private
+
+      def fetch_previous_iteration_state
+        state = Sidekiq.redis { |conn| conn.get("it-#{jid}") }
+
+        if state
+          state_hash = Sidekiq.load_json(state)
+          @executions = state_hash["executions"]
+          @cursor = state_hash["cursor"]
+          @times_interrupted = state_hash["times_interrupted"]
+          @total_time = state_hash["total_time"]
+        end
+      end
+
+      STATE_FLUSH_INTERVAL = 5 # seconds
+      STATE_TTL = 30 * 24 * 60 * 60 # 30 days
+
+      def iterate_with_enumerator(enumerator, arguments)
+        found_record = false
+        state_flushed_at = Time.now
+
+        enumerator.each do |object, cursor|
+          found_record = true
+          around_iteration do
+            each_iteration(object, *arguments)
+          end
+          @cursor = cursor
+          adjust_total_time
+
+          if Time.now - state_flushed_at >= STATE_FLUSH_INTERVAL || job_should_exit?
+            flush_state
+            state_flushed_at = Time.now
+          end
+
+          if job_should_exit?
+            return false
+          end
+        end
+
+        logger.debug("Enumerator found nothing to iterate!") unless found_record
+
+        true
+      end
+
+      def reenqueue_iteration_job
+        @times_interrupted += 1
+        flush_state
+        logger.debug { "Interrupting and re-enqueueing the job (cursor=#{cursor.inspect})" }
+
+        retry_backoff = self.class.get_sidekiq_options.dig("iteration", "retry_backoff") ||
+          Sidekiq.default_configuration.dig("iteration", "retry_backoff")
+
+        # Preserve original jid.
+        self.class.set(jid: jid).perform_in(retry_backoff, *@arguments)
```

Comment:
This will rerun client middleware again. Need to decide, if this is a problem or not.

Reply:
Yeah, I don't like the current approach too.
This is exactly how I implemented it initially, but in the added tests for this PR I now need to rescue this exception manually and do all the rescheduling logic for tests to pass. I will look at if I can somehow better solve this in tests.